csharp/src/Google.Protobuf/CodedOutputWriter.cs

```diff
@@ -0,0 +1,666 @@
+            }
+            else
+            {
+                while (value > 127)
+                {
+                    Ensure(1);
+                    span[buffered++] = (byte)((value & 0x7F) | 0x80);
+                    value >>= 7;
+                }
+
+                Ensure(1);
+                span[buffered++] = (byte)value;
+            }
+        }
+
+        internal void WriteRawLittleEndian32(uint value)
+        {
+            const int length = 4;
+
+            Ensure(length);
+
+            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(buffered), value);
+
+            buffered += length;
+        }
+
+        internal void WriteRawLittleEndian64(ulong value)
+        {
+            const int length = 8;
+
+            Ensure(length);
+
+            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(buffered), value);
+
+            buffered += length;
+        }
+
+        internal void WriteRawByte(byte value)
+        {
+            if (buffered > span.Length)
+            {
+                EnsureMore(1);
+            }
+
+            span[buffered++] = value;
+        }
+
+        /// <summary>
+        /// Writes out an array of bytes.
+        /// </summary>
+        internal void WriteRawBytes(byte[] value)
+        {
+            WriteRawBytes(value, 0, value.Length);
+        }
+
+        /// <summary>
+        /// Writes out part of an array of bytes.
+        /// </summary>
+        internal void WriteRawBytes(byte[] value, int offset, int length)
+        {
+            Write(value.AsSpan(offset, length));
+        }
+
+        #endregion
+
+        #region Buffer
+        /// <summary>
+        /// Copies the caller's buffer into the writer.
+        /// </summary>
+        /// <param name="source">The buffer to copy in.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Write(ReadOnlySpan<byte> source)
+        {
+            if (span.Length - buffered >= source.Length)
+            {
+                source.CopyTo(span.Slice(buffered));
+                buffered += source.Length;
+            }
+            else
+            {
+                WriteMultiBuffer(source);
+            }
+        }
+
+        /// <summary>
+        /// Acquires a new buffer if necessary to ensure that some given number of bytes can be written to a single buffer.
+        /// </summary>
+        /// <param name="count">The number of bytes that must be allocated in a single buffer.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Ensure(int count = 1)
+        {
+            if (span.Length < count + buffered)
+            {
+                EnsureMore(count);
+            }
+        }
+
+        /// <summary>
+        /// Gets a fresh span to write to, with an optional minimum size.
+        /// </summary>
+        /// <param name="count">The minimum size for the next requested buffer.</param>
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
just curious why NoInlining was used here?

Reply:
I've added a comment

> We don't want this method inlined because it is called infrequently.
> It should only be called when the end of the current span is reached.